### PR TITLE
Refactor addon registration and installation

### PR DESCRIPTION
**Pull request by graupnerdemos**
Installing and loading Addons in demosplan is a tedious multistep process.
Some actions happen on the file system, some in Composer, some in Symfony.
Of the actions in Symfony, some need to occur during the container compile
step, some always and some after.

Previously, all of this was more or less mangled into the AddonRegistry
and the AddonInstallFromZipCommand. This commit is a step towards cleaning
things up and making the process a little more understandable and in
consequence hopefully also more maintainable.

As of now, to install an addon zip bundle (installation via pure
composer package is still not fully implemented), the AddonInstallFromZipCommand
has the logic for

- unarchiving the zip to the addons cache (which, as a reminder, is only
  responsible for storing zipped addons and will not be utilized for
  composer-only addons)
- handling the composer "installation" aka the local package registration
  and dependency resolving of the addon
- leaving the system bootable with prepared autoloading for addons

it then passes on to the Registrator which

- takes care of ensuring the addon is valid (not very comprehensive
  yet but that shall be extended upon in due time)
- loads and stores the addon's manifest into the addons.yaml

next, the AddonBuildFrontendCommand takes over and

- leaves again if there is no frontend configuration necessary, or

- installs the required frontend dependencies with Yarn
- runs the production build for the frontend assets assuming the
  addon was packaged using @demos-europe/demosplan-addon

Furthermore, as previously the autoloading, the bundle loading
was extracted into a standalone system.

The AddonRegistry is now an actually runtime usable registry
of all availble addons and configured during container compilation
to receive the AddonInfo instances which replace the AddonInfoProvider
system.

### How to review/test

Install the demospipes addon in it's current state, it should install including frontend stuffs.

### CODE REVIEW

File: demosplan/DemosPlanCoreBundle/Addon/AddonRegistry.php [hunk 1]
```diff
@@ -12,205 +12,104 @@
...
+    public function boot(array $addonInfos = [])
     {
-        if (file_exists(DemosPlanPath::getRootPath('addons/addons.yaml'))) {
-            $configFile = Yaml::parseFile(DemosPlanPath::getRootPath('addons/addons.yaml'));
-            if (is_array($configFile) && array_key_exists('addons', $configFile) && is_array($configFile['addons'])) {
-                $this->addons = \collect($configFile['addons']);
-            }
+        if ([] !== $this->addonInfos) {
+            // TODO: throw exception? go into existential crisis? are dinosaurs even extinct?
+            // no they're not, have you heard about chickens?
```
Comment (AlexMenschoid) [inline demosplan/DemosPlanCoreBundle/Addon/AddonRegistry.php:43]: I think this should be done and not left as a TODO.

File: demosplan/DemosPlanCoreBundle/Addon/Registrator.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     * @return array<string, array<string, string>>
+     */
+    private function loadManifest(string $addonName): array
+    {
+        try {
+            // TODO: Fix manifest parsing
+            /*$config = Yaml::parseFile($this->installedAddons->getManifestPath($addonName));
+            $processor = new Processor();
+            $parsed = $processor->processConfiguration(new ManifestConfiguration(), [$config['demosplan_addon']]);
+            var_export($parsed);*/
```
Comment (AlexMenschoid) [inline demosplan/DemosPlanCoreBundle/Addon/Registrator.php:93]: We should either try to fix it now or remove it as it is not strictly necessary

File: demosplan/DemosPlanCoreBundle/Command/Addon/AddonBuildFrontendCommand.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $output->writeln("Building frontend assets for {$input->getArgument('addon-name')}");
+
+        $addonInfo = $this->registry[$input->getArgument('addon-name')];
+
+        Batch::create($this->getApplication(), $output)
+            ->addShell(['yarn', 'install', '--frozen-lockfile'], $addonInfo->getInstallPath())
+            ->addShell(['yarn', 'run', 'webpack', '--node-env=production'], $addonInfo->getInstallPath())
+            ->run();
```
Comment (AlexMenschoid) [inline demosplan/DemosPlanCoreBundle/Command/Addon/AddonBuildFrontendCommand.php:40]: If yarn is like composer, we should check for the actual return code to catch errors.

**Review by AlexMenschoid — CHANGES_REQUESTED** [review]
I'm going to try to fix my own remarks later today to keep this moving.

File: demosplan/DemosPlanCoreBundle/Addon/AddonRegistry.php (hunk 1 above)
Comment (AlexMenschoid) [reply demosplan/DemosPlanCoreBundle/Addon/AddonRegistry.php]: Fixed in [72ddb80](https://github.com/demos-europe/demosplan-core/pull/640/commits/72ddb800a9cb35f0f573c39780919e9f96e2c094)

File: demosplan/DemosPlanCoreBundle/Addon/Registrator.php (hunk 2 above)
Comment (AlexMenschoid) [reply demosplan/DemosPlanCoreBundle/Addon/Registrator.php]: added in [a6f0cfa](https://github.com/demos-europe/demosplan-core/pull/640/commits/a6f0cfad7075e4deeb474b49a44f62ee83ab780e)

File: demosplan/DemosPlanCoreBundle/Command/Addon/AddonBuildFrontendCommand.php (hunk 3 above)
Comment (AlexMenschoid) [reply demosplan/DemosPlanCoreBundle/Command/Addon/AddonBuildFrontendCommand.php]: added in [8116b4a](https://github.com/demos-europe/demosplan-core/pull/640/commits/8116b4aeaa93898ad42f9c8f16d40f2bc9347f35)

File: demosplan/DemosPlanCoreBundle/Addon/AddonManifestCollection.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace demosplan\DemosPlanCoreBundle\Addon;
+
+use demosplan\DemosPlanCoreBundle\Utilities\DemosPlanPath;
+use Symfony\Component\Yaml\Yaml;
+
+// TODO: find a better name for this!!!
```
Comment (graupnerdemos) [inline demosplan/DemosPlanCoreBundle/Addon/AddonManifestCollection.php:8]: I think we can drop this TODO now, @AlexMenschoid ?

File: demosplan/DemosPlanCoreBundle/Command/Addon/AddonInstallFromZipCommand.php [hunk 5]
```diff
@@ -108,27 +128,17 @@ public function execute(InputInterface $input, OutputInterface $output): int
...
-
-            if (array_key_exists('ui', $packageMeta['manifest'])) {
-                // TODO: fix frontend build
-                /*
-                Batch::create($this->getApplication(), $output)
-                    ->addShell(['yarn', 'install', '--frozen-lockfile'], $packageMeta['install_path'])
-                    ->addShell(['yarn', 'run', 'webpack', '--node-env=production'], $packageMeta['install_path'])
-                    ->run();*/
-            }
+            $activeProject = $this->getApplication()->getKernel()->getActiveProject();
```
[the remark this answers is not in the document]
Comment (graupnerdemos) [reply demosplan/DemosPlanCoreBundle/Command/Addon/AddonInstallFromZipCommand.php:134]: yes, but no, phpstan. you're just not able to read the correct type.

File: demosplan/DemosPlanCoreBundle/Addon/AddonManifestCollection.php (hunk 4 above)
Comment (AlexMenschoid) [reply demosplan/DemosPlanCoreBundle/Addon/AddonManifestCollection.php]: done in [4b4cdd8](https://github.com/demos-europe/demosplan-core/pull/640/commits/4b4cdd8e956919fa555529ec615ce9b1c459f421)

**Comment by AlexMenschoid** [comment]
I also added missing license headers to all the new files in [59567ee](https://github.com/demos-europe/demosplan-core/pull/640/commits/59567ee44d320edaa68f9f6dd8ec76753c2b824a)

**Review by AlexMenschoid — APPROVED** [review]
I'm approving this for myself and @graupnerdemos. We discussed this and he reviewed my parts and vice versa.

**Pull request merged**